File: src/agent_run/domain.py
"""Frozen types shared by configuration, adapters, and storage."""

from __future__ import annotations

import math
import secrets
from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from pathlib import Path
from types import MappingProxyType
from typing import Final, Mapping, NewType

from .errors import StateTransitionError, ValidationError


AgentId = NewType("AgentId", str)
_MAX_EXTERNAL_ID_LENGTH = 512
_LOWER_HEX = frozenset("0123456789abcdef")


class AgentStatus(str, Enum):
    CREATED = "created"
    STARTING = "starting"
    RUNNING = "running"
    CANCELLING = "cancelling"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    TIMED_OUT = "timed_out"
    CANCELLED = "cancelled"
    LOST = "lost"


ACTIVE: Final = frozenset(
    {
        AgentStatus.CREATED,
        AgentStatus.STARTING,
        AgentStatus.RUNNING,
        AgentStatus.CANCELLING,
    }
)
TERMINAL: Final = frozenset(
    {
        AgentStatus.SUCCEEDED,
        AgentStatus.FAILED,
        AgentStatus.TIMED_OUT,
        AgentStatus.CANCELLED,
        AgentStatus.LOST,
    }
)
TRANSITIONS: Final[Mapping[AgentStatus, frozenset[AgentStatus]]] = MappingProxyType(
    {
        AgentStatus.CREATED: frozenset(
            {
                AgentStatus.STARTING,
                AgentStatus.CANCELLED,
                AgentStatus.FAILED,
                AgentStatus.LOST,
            }
        ),
        AgentStatus.STARTING: frozenset(
            {
                AgentStatus.RUNNING,
                AgentStatus.CANCELLED,
                AgentStatus.FAILED,
                AgentStatus.LOST,
            }
        ),
        AgentStatus.RUNNING: frozenset(
            {
                AgentStatus.SUCCEEDED,
                AgentStatus.FAILED,
                AgentStatus.TIMED_OUT,
                AgentStatus.CANCELLING,
                AgentStatus.LOST,
            }
        ),
        AgentStatus.CANCELLING: frozenset(
            {AgentStatus.CANCELLED, AgentStatus.LOST}
        ),
        **{status: frozenset() for status in TERMINAL},
    }
)


def validate_transition(current: AgentStatus, target: AgentStatus) -> None:
    if (
        not isinstance(current, AgentStatus)
        or not isinstance(target, AgentStatus)
        or target not in TRANSITIONS.get(current, ())
    ):
        raise StateTransitionError(f"invalid agent transition: {current} -> {target}")


def validate_agent_id(value: str) -> AgentId:
    if not isinstance(value, str) or len(value) != 29:
        raise ValidationError("agent_id must match ag-YYYYMMDD-HHMMSS-<10 lowercase hex>")
    if not value.startswith("ag-") or value[18] != "-":
        raise ValidationError("agent_id must match ag-YYYYMMDD-HHMMSS-<10 lowercase hex>")
    try:
        datetime.strptime(value[3:18], "%Y%m%d-%H%M%S")
    except ValueError as error:
        raise ValidationError(
            "agent_id must match ag-YYYYMMDD-HHMMSS-<10 lowercase hex>"
        ) from error
    if any(character not in _LOWER_HEX for character in value[19:]):
        raise ValidationError("agent_id must match ag-YYYYMMDD-HHMMSS-<10 lowercase hex>")
    return AgentId(value)


def new_agent_id() -> AgentId:
    value = f"ag-{datetime.now(timezone.utc):%Y%m%d-%H%M%S}-{secrets.token_hex(5)}"
    return validate_agent_id(value)


def _nonblank(name: str, value: str) -> None:
    if not isinstance(value, str) or not value.strip():
        raise ValidationError(f"{name} must be a nonblank string")


def _bounded_id(name: str, value: str) -> None:
    _nonblank(name, value)
    if len(value) > _MAX_EXTERNAL_ID_LENGTH:
        raise ValidationError(f"{name} must be at most {_MAX_EXTERNAL_ID_LENGTH} characters")


def _existing_directory(name: str, value: Path) -> Path:
    if not isinstance(value, Path) or not value.is_absolute():
        raise ValidationError(f"{name} must be an absolute existing directory")
    try:
        resolved = value.resolve(strict=True)
    except (OSError, RuntimeError) as error:
        raise ValidationError(f"{name} must be an absolute existing directory") from error
    if not resolved.is_dir():
        raise ValidationError(f"{name} must be an absolute existing directory")
    return resolved


@dataclass(frozen=True)
class OrchestratorRef:
    transport: str
    external_session_id: str
    external_turn_id: str | None = None

    def __post_init__(self) -> None:
        _bounded_id("transport", self.transport)
        _bounded_id("external_session_id", self.external_session_id)
        if self.external_turn_id is not None:
            _bounded_id("external_turn_id", self.external_turn_id)


@dataclass(frozen=True)
class StartRequest:
    runtime: str
    model: str
    profile: str
    task: str
    workdir: Path
    write: bool = False
    effort: str | None = None
    timeout_seconds: float | None = None
    read_roots: tuple[Path, ...] = ()
    output_schema: dict | None = None
    orchestrator: OrchestratorRef | None = None
    request_id: str | None = None
    fast: bool = False

    def __post_init__(self) -> None:
        for name in ("runtime", "model", "profile", "task"):
            _nonblank(name, getattr(self, name))
        if not isinstance(self.fast, bool):
            raise ValidationError("fast must be a boolean")
        if self.timeout_seconds is not None:
            if isinstance(self.timeout_seconds, bool) or not isinstance(
                self.timeout_seconds, (int, float)
            ):
                raise ValidationError("timeout_seconds must be positive and finite")
            if self.timeout_seconds <= 0 or not math.isfinite(self.timeout_seconds):
                raise ValidationError("timeout_seconds must be positive and finite")
        if not isinstance(self.read_roots, tuple):
            raise ValidationError("read_roots must be a tuple")
        workdir = _existing_directory("workdir", self.workdir)
        read_roots = tuple(
            _existing_directory(f"read_roots[{index}]", path)
            for index, path in enumerate(self.read_roots)
        )
        if len(set(read_roots)) != len(read_roots):
            raise ValidationError("read_roots must not contain duplicates")
        if self.effort is not None:
            _nonblank("effort", self.effort)
        if self.output_schema is not None and not isinstance(self.output_schema, dict):
            raise ValidationError("output_schema must be a dict or None")
        if self.orchestrator is not None and not isinstance(
            self.orchestrator, OrchestratorRef
        ):
            raise ValidationError("orchestrator must be an OrchestratorRef or None")
        if self.request_id is not None:
            _bounded_id("request_id", self.request_id)
        object.__setattr__(self, "workdir", workdir)
        object.__setattr__(self, "read_roots", read_roots)



class MessageRole(str, Enum):
    USER = "user"
    ASSISTANT = "assistant"
    TOOL_CALL = "tool_call"
    TOOL_RESULT = "tool_result"
    SYSTEM = "system"


@dataclass(frozen=True, slots=True)
class Message:
    at: float
    role: MessageRole
    content: str
    name: str | None = None
    raw_ref: str | None = None

    def __post_init__(self) -> None:
        if isinstance(self.at, bool) or not isinstance(self.at, (int, float)):
            raise ValidationError("message at must be finite and nonnegative")
        if self.at < 0 or not math.isfinite(self.at):
            raise ValidationError("message at must be finite and nonnegative")
        if not isinstance(self.role, MessageRole):
            raise ValidationError("message role must be a MessageRole")
        _nonblank("message content", self.content)


@dataclass(frozen=True, slots=True)
class Outcome:
    status: AgentStatus
    exit_code: int | None = None
    failure_kind: str | None = None
    failure_text: str | None = None
    runtime_session_id: str | None = None
    answer_path: Path | None = None
    answer_bytes: int | None = None
    answer_sha256: str | None = None

    def __post_init__(self) -> None:
        if not isinstance(self.status, AgentStatus) or self.status not in TERMINAL:
            raise ValidationError("outcome status must be terminal")
        if self.answer_bytes is not None and (
            isinstance(self.answer_bytes, bool)
            or not isinstance(self.answer_bytes, int)
            or self.answer_bytes < 0
        ):
            raise ValidationError("answer_bytes must be a nonnegative integer or None")

File: crates/agent-run-domain/src/domain.rs
use crate::{error::invalid, Error, Result};
use chrono::{NaiveDateTime, Utc};
use serde::{Deserialize, Serialize};
use std::{collections::BTreeSet, fmt, path::PathBuf, str::FromStr};

/// Constraints that a resolved execution policy can require or evidence.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Constraint {
    WebToolsDisabled,
    ExternalNetworkIsolation,
    LoopbackTcpIsolation,
    UnixIpcIsolation,
    McpIpcIsolation,
    FilesystemWriteIsolation,
    FilesystemReadIsolation,
    PluginImmutability,
}
impl Constraint {
    /// Lists every recognized constraint in stable declaration order.
    pub const ALL: [Self; 8] = [
        Self::WebToolsDisabled,
        Self::ExternalNetworkIsolation,
        Self::LoopbackTcpIsolation,
        Self::UnixIpcIsolation,
        Self::McpIpcIsolation,
        Self::FilesystemWriteIsolation,
        Self::FilesystemReadIsolation,
        Self::PluginImmutability,
    ];
}

pub fn now() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize)]
#[serde(transparent)]
pub struct AgentId(String);
impl AgentId {
    pub fn new() -> Self {
        Self(format!(
            "ag-{}-{}",
            Utc::now().format("%Y%m%d-%H%M%S"),
            &uuid::Uuid::new_v4().simple().to_string()[..10]
        ))
    }
    pub fn as_str(&self) -> &str {
        &self.0
    }
}
impl Default for AgentId {
    fn default() -> Self {
        Self::new()
    }
}
impl fmt::Display for AgentId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}
impl FromStr for AgentId {
    type Err = Error;
    fn from_str(s: &str) -> Result<Self> {
        if s.len() != 29
            || !s.is_ascii()
            || !s.starts_with("ag-")
            || s.as_bytes()[18] != b'-'
            || !s[19..]
                .bytes()
                .all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
            || NaiveDateTime::parse_from_str(&s[3..18], "%Y%m%d-%H%M%S").is_err()
        {
            return Err(invalid(
                "agent_id must match ag-YYYYMMDD-HHMMSS-<10 lowercase hex>",
            ));
        }
        Ok(Self(s.into()))
    }
}
impl<'de> Deserialize<'de> for AgentId {
    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> std::result::Result<Self, D::Error> {
        String::deserialize(d)?
            .parse()
            .map_err(serde::de::Error::custom)
    }
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Status {
    Created,
    Starting,
    Running,
    Cancelling,
    Succeeded,
    Failed,
    TimedOut,
    Cancelled,
    Lost,
}
impl Status {
    pub const ALL: [Self; 9] = [
        Self::Created,
        Self::Starting,
        Self::Running,
        Self::Cancelling,
        Self::Succeeded,
        Self::Failed,
        Self::TimedOut,
        Self::Cancelled,
        Self::Lost,
    ];
    pub fn terminal(self) -> bool {
        matches!(
            self,
            Self::Succeeded | Self::Failed | Self::TimedOut | Self::Cancelled | Self::Lost
        )
    }
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Created => "created",
            Self::Starting => "starting",
            Self::Running => "running",
            Self::Cancelling => "cancelling",
            Self::Succeeded => "succeeded",
            Self::Failed => "failed",
            Self::TimedOut => "timed_out",
            Self::Cancelled => "cancelled",
            Self::Lost => "lost",
        }
    }
    pub fn transition(self, to: Self) -> Result<()> {
        let ok = match self {
            Self::Created => matches!(
                to,
                Self::Starting | Self::Cancelled | Self::Failed | Self::Lost
            ),
            Self::Starting => matches!(
                to,
                Self::Running | Self::Cancelled | Self::Failed | Self::Lost
            ),
            Self::Running => matches!(
                to,
                Self::Succeeded | Self::Failed | Self::TimedOut | Self::Cancelling | Self::Lost
            ),
            Self::Cancelling => matches!(to, Self::Cancelled | Self::Lost),
            _ => false,
        };
        if ok {
            Ok(())
        } else {
            Err(Error::Transition(format!(
                "{} -> {}",
                self.as_str(),
                to.as_str()
            )))
        }
    }
}
impl FromStr for Status {
    type Err = Error;
    fn from_str(v: &str) -> Result<Self> {
        Self::ALL
            .into_iter()
            .find(|s| s.as_str() == v)
            .ok_or_else(|| invalid("unknown status"))
    }
}

pub fn nonblank(label: &str, s: &str) -> Result<()> {
    if s.trim().is_empty() || s.contains('\0') {
        return Err(invalid(format!(
            "{label} must be a nonblank NUL-free string"
        )));
    }
    Ok(())
}
pub fn external_id(label: &str, s: &str) -> Result<()> {
    nonblank(label, s)?;
    if s.chars().count() > 512 {
        return Err(invalid(format!("{label} exceeds 512 characters")));
    }
    Ok(())
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct OrchestratorRef {
    pub transport: String,
    pub external_session_id: String,
    #[serde(default)]
    pub external_turn_id: Option<String>,
}
impl OrchestratorRef {
    pub fn validate(&self) -> Result<()> {
        external_id("transport", &self.transport)?;
        external_id("external_session_id", &self.external_session_id)?;
        if let Some(v) = &self.external_turn_id {
            external_id("external_turn_id", v)?;
        }
        Ok(())
    }
}
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct StartRequest {
    pub runtime: String,
    pub model: String,
    pub profile: String,
    pub task: String,
    pub workdir: PathBuf,
    #[serde(default)]
    pub write: bool,
    #[serde(default)]
    pub fast: bool,
    #[serde(default)]
    pub effort: Option<String>,
    /// Legacy metadata only. Never an execution deadline.
    #[serde(default)]
    pub timeout_seconds: Option<f64>,
    #[serde(default)]
    pub read_roots: Vec<PathBuf>,
    #[serde(default)]
    pub output_schema: Option<serde_json::Map<String, serde_json::Value>>,
    #[serde(default)]
    pub orchestrator: Option<OrchestratorRef>,
    #[serde(default)]
    pub request_id: Option<String>,
    #[serde(default)]
    pub account: Option<String>,
    #[serde(default, deserialize_with = "unique_constraints")]
    pub required_constraints: BTreeSet<Constraint>,
}
fn unique_constraints<'de, D: serde::Deserializer<'de>>(
    d: D,
) -> std::result::Result<BTreeSet<Constraint>, D::Error> {
    let a = Vec::<Constraint>::deserialize(d)?;
    let b: BTreeSet<_> = a.iter().copied().collect();
    if a.len() != b.len() {
        return Err(serde::de::Error::custom("duplicate constraints"));
    }
    Ok(b)
}
impl StartRequest {
    pub fn validate(&mut self) -> Result<()> {
        for (name, s) in [
            ("runtime", &self.runtime),
            ("model", &self.model),
            ("profile", &self.profile),
            ("task", &self.task),
        ] {
            nonblank(name, s)?;
        }
        if self.task.len() > 512 * 1024 {
            return Err(invalid("task exceeds 512 KiB"));
        }
        for (name, s) in [("effort", &self.effort), ("account", &self.account)] {
            if let Some(s) = s {
                nonblank(name, s)?;
            }
        }
        if let Some(v) = self.timeout_seconds {
            if !v.is_finite() || v <= 0.0 {
                return Err(invalid("timeout_seconds must be positive and finite"));
            }
        }
        self.workdir = existing_dir(&self.workdir)?;
        self.read_roots = self
            .read_roots
            .iter()
            .map(|p| existing_dir(p))
            .collect::<Result<_>>()?;
        let mut seen = BTreeSet::new();
        if self.read_roots.iter().any(|p| !seen.insert(p.clone())) {
            return Err(invalid("read_roots must not contain duplicates"));
        }
        if let Some(r) = &self.orchestrator {
            r.validate()?;
        }
        if let Some(r) = &self.request_id {
            external_id("request_id", r)?;
        }
        Ok(())
    }
}
pub fn existing_dir(path: &std::path::Path) -> Result<PathBuf> {
    if !path.is_absolute() {
        return Err(invalid("path must be an absolute existing directory"));
    }
    let p = path
        .canonicalize()
        .map_err(|_| invalid("directory does not exist"))?;
    if !p.is_dir() {
        return Err(invalid("path is not a directory"));
    }
    Ok(p)
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Outcome {
    pub status: Status,
    pub exit_code: Option<i32>,
    pub failure_kind: Option<String>,
    pub failure_text: Option<String>,
    pub runtime_session_id: Option<String>,
}
impl Outcome {
    pub fn success(session: Option<String>) -> Self {
        Self {
            status: Status::Succeeded,
            exit_code: Some(0),
            failure_kind: None,
            failure_text: None,
            runtime_session_id: session,
        }
    }
    pub fn failure(kind: impl Into<String>) -> Self {
        Self {
            status: Status::Failed,
            exit_code: None,
            failure_kind: Some(kind.into()),
            failure_text: None,
            runtime_session_id: None,
        }
    }
}
